drivers/overlay/mount.go

```diff
@@ -77,12 +79,111 @@ func mountFromMain() {
 		fatal(err)
 	}
 
-	if err := os.Chdir(flag.Arg(0)); err != nil {
+	// Mount the arguments passed from the specified directory. Some of the
+	// paths mentioned in the values we pass to the kernel are relative to
+	// the specified directory.
+	homedir := flag.Arg(0)
+	if err := os.Chdir(homedir); err != nil {
 		fatal(err)
 	}
 
-	if err := unix.Mount(options.Device, options.Target, options.Type, uintptr(options.Flag), options.Label); err != nil {
-		fatal(err)
+	pageSize := unix.Getpagesize()
+	if len(options.Label) < pageSize {
+		if err := unix.Mount(options.Device, options.Target, options.Type, uintptr(options.Flag), options.Label); err != nil {
+			fatal(err)
+		}
+		os.Exit(0)
+	}
+
+	// Those arguments still took up too much space.  Open the diff
+	// directories and use their descriptor numbers as lowers, using
+	// /proc/self/fd as the current directory.
+
+	// Split out the various options, since we need to manipulate the
+	// paths, but we don't want to mess with other options.
+	var upper, work, lower, label, other string
+	for _, arg := range strings.Split(options.Label, ",") {
+		if strings.HasPrefix(arg, "upperdir=") {
+			upper = strings.TrimPrefix(arg, "upperdir=")
+		} else if strings.HasPrefix(arg, "workdir=") {
+			work = strings.TrimPrefix(arg, "workdir=")
+		} else if strings.HasPrefix(arg, "lowerdir=") {
+			lower = strings.TrimPrefix(arg, "lowerdir=")
+		} else if strings.HasPrefix(arg, "label=") {
+			label = arg
+		} else {
+			if other == "" {
+				other = arg
+			} else {
+				other = other + "," + arg
+			}
+		}
+	}
+
+	// Don't get tricked if there are no lowers.
+	lowers := strings.Split(lower, ":")
+	if len(lowers) == 1 && lowers[0] == "" {
+		lowers = []string{}
+	}
+
+	// Make sure upperdir, workdir, and the target are absolute paths.
+	if upper != "" {
+		if !filepath.IsAbs(upper) {
+			upper = filepath.Join(homedir, upper)
+		}
+		upper = "upperdir=" + upper
+	}
+	if work != "" {
+		if !filepath.IsAbs(work) {
+			work = filepath.Join(homedir, work)
+		}
+		work = "workdir=" + work
+	}
+	if !filepath.IsAbs(options.Target) {
+		options.Target = filepath.Join(homedir, options.Target)
+	}
+
+	// Get a descriptor for each lower, and use that descriptor's name as
+	// the new value for the list of lowers, because it's shorter.
+	if lower != "" {
+		for i := range lowers {
+			lowerFd, err := unix.Open(lowers[i], unix.O_RDONLY, 0)
+			if err != nil {
+				fatal(err)
+			}
+			lowers[i] = fmt.Sprintf("%d", lowerFd)
+		}
+		lower = "lowerdir=" + strings.Join(lowers, ":")
+	}
+
+	// Reconstruct the Label field.
+	options.Label = ""
+	for _, arg := range []string{upper, work, lower, label, other} {
```

Comment:
[A] …

something like
```go
for _, pair := range []struct{optionName, value string} {
   { "upperdir=", upper },
   …
   { "", other },
}
```
would allow most of this function to only deal with the individual values instead of the mount syntax.

---

(Ultimately, this “turn option values into a mount option syntax string” code might be useful to completely factor out…)